### PR TITLE
[DataGrid] Make `checkboxSelection` respect the `disableMultipleRowSelection` prop

### PR DESCRIPTION
**Pull request by cherniavskii**
Closes https://github.com/mui/mui-x/issues/5286

Preview: https://deploy-preview-11448--material-ui-x.netlify.app/x/react-data-grid/row-selection/#checkbox-selection

### CODE REVIEW

File: packages/grid/x-data-grid/src/hooks/features/rowSelection/utils.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    DataGridProcessedProps,
+    'signature' | 'disableMultipleRowSelection' | 'checkboxSelection'
+  >,
+) {
+  if (props.signature === GridSignature.DataGrid) {
+    // DataGrid Community has multiple row selection enabled only if checkbox selection is enabled.
+    return props.checkboxSelection && props.disableMultipleRowSelection !== true;
+  }
+  return !props.disableMultipleRowSelection;
+}
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/hooks/features/rowSelection/utils.ts:15]: This is a bit messy, but the main change in this PR is that you can now pass the `disableMultipleRowSelection` prop to MIT `DataGrid`.
The default behavior hasn't changed, so this PR does not bring any breaking changes.

File: packages/grid/x-data-grid/src/hooks/features/rowSelection/utils.ts
Comment (MBilalShafi) [inline packages/grid/x-data-grid/src/hooks/features/rowSelection/utils.ts:1]: Will it make sense to add a toggle button to toggle between `checkboxSelection` and normal selection in the Single Row Selection demo (now that both can work in a similar way)

```diff
diff --git a/docs/data/data-grid/row-selection/SingleRowSelectionGrid.tsx b/docs/data/data-grid/row-selection/SingleRowSelectionGrid.tsx
index c5446d057..417128b90 100644
--- a/docs/data/data-grid/row-selection/SingleRowSelectionGrid.tsx
+++ b/docs/data/data-grid/row-selection/SingleRowSelectionGrid.tsx
@@ -1,8 +1,11 @@
 import * as React from 'react';
+import Button from '@mui/material/Button';
 import { DataGrid } from '@mui/x-data-grid';
 import { useDemoData } from '@mui/x-data-grid-generator';
 
 export default function SingleRowSelectionGrid() {
+  const [checkboxSelection, setCheckboxSelection] = React.useState(false);
+
   const { data } = useDemoData({
     dataSet: 'Commodity',
     rowLength: 10,
@@ -11,7 +14,17 @@ export default function SingleRowSelectionGrid() {
 
   return (
     <div style={{ height: 400, width: '100%' }}>
-      <DataGrid {...data} />
+      <Button
+        sx={{ mb: 1 }}
+        onClick={() => setCheckboxSelection(!checkboxSelection)}
+      >
+        Toggle checkbox selection
+      </Button>
+      <DataGrid
+        checkboxSelection={checkboxSelection}
+        disableMultipleRowSelection
+        {...data}
+      />
     </div>
   );
 }
```

File: packages/grid/x-data-grid/src/DataGrid/DataGrid.tsx [hunk 2]
```diff
@@ -165,6 +165,12 @@ DataGridRaw.propTypes = {
    * @ignore - do not document
    */
   disableEval: PropTypes.bool,
+  /**
+   * If `true`, multiple selection using the Ctrl/CMD or Shift key is disabled.
+   * The MIT DataGrid will ignore this prop, unless `checkboxSelection` is enabled.
+   * @default false (`true` for MIT DataGrid)
```
Comment (MBilalShafi) [inline packages/grid/x-data-grid/src/DataGrid/DataGrid.tsx:171]: > (`true` for MIT DataGrid)

If that is the case, shouldn't [this demo](https://deploy-preview-11448--material-ui-x.netlify.app/x/react-data-grid/row-selection/#checkbox-selection) disallow multiple selection?
Seems like it's false for MIT Data Grid too. 🤔

File: docs/translations/api-docs/data-grid/data-grid.json
Comment (MBilalShafi) [inline docs/translations/api-docs/data-grid/data-grid.json:1]: Nitpick unrelated to this PR.
```diff
diff --git a/docs/data/data-grid/row-selection/row-selection.md b/docs/data/data-grid/row-selection/row-selection.md
index 3cc10e15d..2f1297175 100644
--- a/docs/data/data-grid/row-selection/row-selection.md
+++ b/docs/data/data-grid/row-selection/row-selection.md
@@ -14,7 +14,7 @@ To unselect a row, hold the <kbd class="key">Ctrl</kbd> key and click on it.
 
 On the `DataGridPro` component, you can select multiple rows in two ways:
 
-- To select multiple independent rows, hold the <kbd class="key">Ctrl</kbd> key while selecting rows.
+- To select multiple independent rows, hold the <kbd class="key">Ctrl</kbd> (<kbd class="key">Cmd</kbd> on Mac OS) key while selecting rows.
 - To select a range of rows, hold the <kbd class="key">SHIFT</kbd> key while selecting rows.
 - To disable multiple row selection, use `disableMultipleRowSelection={true}`.
```

**Review by MBilalShafi — COMMENTED** [review]
Feels more natural now 👍

File: packages/grid/x-data-grid/src/DataGrid/DataGrid.tsx (hunk 2 above)
Comment (cherniavskii) [reply packages/grid/x-data-grid/src/DataGrid/DataGrid.tsx:171]: The default value for MIT Data Grid is:
- `false` if `checkboxSelection={true}`
- `true` if `checkboxSelection={false}`

I'll update the JSDoc to clarify this:

```
@default false (`true` for MIT DataGrid with `checkboxSelection={false}`)
```

File: packages/grid/x-data-grid/src/hooks/features/rowSelection/utils.ts
Comment (cherniavskii) [reply packages/grid/x-data-grid/src/hooks/features/rowSelection/utils.ts:1]: I would rather keep this specific demo as simple as possible, but I'll this to the checkbox selection demo

Comment (MBilalShafi) [reply packages/grid/x-data-grid/src/hooks/features/rowSelection/utils.ts:1]: Looks even nicer 👍

**Review by MBilalShafi — APPROVED** [review]

File: packages/grid/x-data-grid/src/DataGrid/DataGrid.tsx [hunk 3]
```diff
@@ -165,6 +165,12 @@ DataGridRaw.propTypes = {
    * @ignore - do not document
    */
   disableEval: PropTypes.bool,
+  /**
+   * If `true`, multiple selection using the Ctrl/CMD or Shift key is disabled.
+   * The MIT DataGrid will ignore this prop, unless `checkboxSelection` is enabled.
+   * @default false (`true` MIT DataGrid with `checkboxSelection={false}`)
```
[the remark this answers is not in the document]
Comment (cherniavskii) [reply packages/grid/x-data-grid/src/DataGrid/DataGrid.tsx:171]: Thanks!

**Comment by Ciframa** [comment]
Does it work with Data-Grid-**Pro** for u guys?

**Comment by cherniavskii** [comment]
@Ciframa Sure, here's a demo: https://stackblitz.com/edit/react-bvyzcm

**Pull request merged**